### ISSUE TITLE
CUDA Being Initiated Early After Importing Certain Libraries on Nightly Build

### ISSUE THREAD

**Issue by KumoLiu** [pytorch/pytorch#122085]
### 🐛 Describe the bug

Thank you for all the dedication and effort put into this influential library.
I'm part of a team at NVIDIA, and we encountered some issues when intending to use the PyTorch Nightly version. We noticed that CUDA is initiated early after importing certain libraries, something that differs from the 2.2.1 version behavior which we are used to. It interferes with how our product operates.
Here's a list with a few libraries that initiate CUDA on import:
```
packages = [
    "torchvision",
    "torch._dynamo",
    "torch.distributed._tensor",
    "torch._functorch._aot_autograd.jit_compile_runtime_wrappers",
    "torch._functorch._aot_autograd.dispatch_and_compile_graph",
    "torch._functorch.aot_autograd",
    "torch.distributed._functional_collectives_impl",
    "torch.distributed._functional_collectives",
    "torch._higher_order_ops.map",
]
```
Following the import of these libraries, trying to set CUDA_VISIBLE_DEVICES fails. Here's an example code snippet:
```
import os
import torch._higher_order_ops.map
os.environ['CUDA_VISIBLE_DEVICES'] = '0'

torch.cuda.device_count()  # This does not return 1 as expected
```
Could you look into this situation, or provide some information concerning when we might anticipate a solution?
Your help is much appreciated.

### Versions

Collecting environment information...
PyTorch version: 2.4.0.dev20240317+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.27.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-92-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.128
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090 Ti
GPU 1: NVIDIA GeForce RTX 3090 Ti

Nvidia driver version: 535.154.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             20
On-line CPU(s) list:                0-19
Vendor ID:                          GenuineIntel
Model name:                         12th Gen Intel(R) Core(TM) i7-12700K
CPU family:                         6
Model:                              151
Thread(s) per core:                 2
Core(s) per socket:                 12
Socket(s):                          1
Stepping:                           2
CPU max MHz:                        5000.0000
CPU min MHz:                        800.0000
BogoMIPS:                           7219.20
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          512 KiB (12 instances)
L1i cache:                          512 KiB (12 instances)
L2 cache:                           12 MiB (9 instances)
L3 cache:                           25 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-19
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] flake8-bugbear==24.2.6
[pip3] flake8-comprehensions==3.14.0
[pip3] mypy==1.9.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.22.2
[pip3] onnx==1.14.0
[pip3] onnxruntime==1.17.1
[pip3] pytorch-ignite==0.4.11
[pip3] pytorch-quantization==2.1.2
[pip3] pytorch-triton==3.0.0+989adb9a29
[pip3] torch==2.4.0.dev20240317+cu121
[pip3] torch-tensorrt==2.0.0.dev0
[pip3] torchaudio==2.2.0.dev20240317+cu121
[pip3] torchdata==0.7.0a0
[pip3] torchtext==0.16.0a0
[pip3] torchvision==0.18.0.dev20240317+cu121
[pip3] triton==2.2.0
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @kadeng @ptrblck @msaroufim @bdhirsh @anijain2305 @chauhang

**Comment by bdhirsh** [comment]
Marking with the pt2 oncall since most of the relevant files are PT2 related

**Comment by ezyang** [comment]
We should not be initializing CUDA upon import.

**Comment by Nic-Ma** [comment]
Hi @ezyang ,

Thanks for your quick response here.
Could you please help connect the related engineer to help fix it?
We are updating the latest PyTorch for our products https://github.com/Project-MONAI/MONAI and https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara/containers/monai-toolkit.
It's a blocker for the release.

Thanks in advance.

**Comment by Chillee** [comment]
Could it be this? https://github.com/pytorch/pytorch/pull/121864

**Comment by wyli** [comment]
looking into this further, I think it's related to the `device_count` caching issue: https://github.com/pytorch/pytorch/issues/95073,

more recently this module-level device_count call seems to make the issue more noticeable: https://github.com/pytorch/pytorch/blob/4c70ab26ef0a596d779ec38e9bc9c1bfc4373f93/torch/_dynamo/device_interface.py#L198

**Comment by ptrblck** [comment]
@ezyang 
> We should not be initializing CUDA upon import.

Yes, totally agree, but I also don't see a `cuInit` here (unless I'm using a wrong PyTorch build).

We've recently seen a real CUDA init call while importing PyTorch in https://github.com/pytorch/pytorch/issues/116276
and worked on a unit test making sure this won't happen again.
CC @malfet, @Aidyn-A, @atalman as the test laded in https://github.com/pytorch/pytorch/pull/117043 

Running the script from this issue in a debugger also does not show any `cuInit` calls:
```python
lldb --batch -o "b cuInit" -o "run" -- python3 -c "import torch;print(torch.cuda.device_count() > 0); import torch._higher_order_ops.map"
```
while adding `; torch.randn(1).cuda()` indeed breaks.

The script tries to set `CUDA_VISIBLE_DEVICES` after `torch` was imported which is too late.
This env variable should be set before importing any CUDA-enabled libraries in the script or exported in your terminal.

**Comment by wyli** [comment]
sure, I can have a look soon @ezyang

**Comment by wyli** [comment]
put a possible fix here in case it's useful https://github.com/pytorch/pytorch/pull/122795

**Comment by ezyang** [comment]
Actually, reading over our code, this may be an entirely PyTorch inflicted problem

```
DeviceIndex device_count() noexcept {
  // initialize number of devices only once
  static int count = []() {
    try {
      auto result = device_count_impl(/*fail_if_no_driver=*/false);
      TORCH_INTERNAL_ASSERT(
          result <= std::numeric_limits<DeviceIndex>::max(),
          "Too many CUDA devices, DeviceIndex overflowed");
      return result;
    } catch (const c10::Error& ex) {
      // We don't want to fail, but still log the warning
      // msg() returns the message without the stack trace
      TORCH_WARN("CUDA initialization: ", ex.msg());
      return 0;
    }     
  }();    
  return static_cast<DeviceIndex>(count);
}
```

**Comment by ezyang** [comment]
I thought maybe this would work https://github.com/pytorch/pytorch/pull/122805 but it doesn't >:(

```
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (8743ac67)]$ python wu.py
8
8
Traceback (most recent call last):
  File "/data/users/ezyang/b/pytorch/torch/cuda/__init__.py", line 306, in _lazy_init
    queued_call()
  File "/data/users/ezyang/b/pytorch/torch/cuda/__init__.py", line 174, in _check_capability
    capability = get_device_capability(d)
  File "/data/users/ezyang/b/pytorch/torch/cuda/__init__.py", line 430, in get_device_capability
    prop = get_device_properties(device)
  File "/data/users/ezyang/b/pytorch/torch/cuda/__init__.py", line 448, in get_device_properties
    return _get_device_properties(device)  # type: ignore[name-defined]
RuntimeError: device=, num_gpus=

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/data/users/ezyang/b/pytorch/wu.py", line 6, in <module>
    print(torch.empty(10, device='cuda'))
  File "/data/users/ezyang/b/pytorch/torch/cuda/__init__.py", line 312, in _lazy_init
    raise DeferredCudaCallError(msg) from e
torch.cuda.DeferredCudaCallError: CUDA call failed lazily at initialization with error: device=, num_gpus=

CUDA call was originally invoked at:

  File "/data/users/ezyang/b/pytorch/wu.py", line 1, in <module>
    import torch
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/data/users/ezyang/b/pytorch/torch/__init__.py", line 1485, in <module>
    _C._initExtension(manager_path())
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/data/users/ezyang/b/pytorch/torch/cuda/__init__.py", line 238, in <module>
    _lazy_call(_check_capability)
  File "/data/users/ezyang/b/pytorch/torch/cuda/__init__.py", line 235, in _lazy_call
    _queued_calls.append((callable, traceback.format_stack()))
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (8743ac67)]$ cat wu.py
import torch
import os
print(torch.cuda.device_count())
os.environ['CUDA_VISIBLE_DEVICES'] = '0'
print(torch.cuda.device_count())
print(torch.empty(10, device='cuda'))
```

Maybe someone can figure out what I did wrong lol.

**Comment by ezyang** [comment]
Oh, it's because we are using nvml to query device count 🤔

**Comment by ezyang** [comment]
So calling device_count initializes NVML. Is that the root cause of the problem here? Should we stop using NVML to query device here?

**Comment by wyli** [comment]
Hi @ezyang, I think the title of this ticket becomes a bit misleading after all the investigations.

to summarize my understanding of the issue: when importing some torch submodules such as `import torch._dynamo`, somewhat expectedly `torch.cuda.device_count` will be called  https://github.com/pytorch/pytorch/blob/37e3c8f33f165f25184a573183d2aca878b84476/torch/_dynamo/device_interface.py#L198
 and the outcome cached https://github.com/pytorch/pytorch/blob/37e3c8f33f165f25184a573183d2aca878b84476/torch/cuda/__init__.py#L742

So, any setenv CUDA_VISIBLE_DEVICES after the import statement may make the actual device count inconsistent with the (cached) output of `torch.cuda.device_count()`...

**Comment by wyli** [comment]
looking at the implementation https://github.com/pytorch/pytorch/blob/37e3c8f33f165f25184a573183d2aca878b84476/torch/cuda/__init__.py#L748-L749

you're right -- in our use case the `torch.cuda.device_count` calls `_device_count_nvml`. if it runs into `torch._C._cuda_getDeviceCount()`, it indeed weakly init and allocate a small gpu memory.

**Comment by ezyang** [comment]
Oh. So if we get rid of the LRU cache when CUDA is not initialized, that should fix this once and for all. I'll send a patch.

**Comment by wyli** [comment]
agreed, thanks for looking into the issue. 

also this module-level `torch.cuda.device_count` could be delayed as well https://github.com/pytorch/pytorch/blob/37e3c8f33f165f25184a573183d2aca878b84476/torch/_dynamo/device_interface.py#L198

**Comment by ezyang** [comment]
Yeah, your patch is still good, I still want it in

**Comment by youkaichao** [comment]
It's sad the related bug fix is not included in pytorch 2.3.1, and we don't have plan for 2.3.2 . We have been troubled by this for a long time, see https://github.com/vllm-project/vllm/issues/6056 :( Now we need to wait for 2.4 release.
